### PR TITLE
[15.0][MIG] project_stock: Migration to 15.0

### PR DESCRIPTION
**Pull request by victoralmau**
Migration to 15.0

Please @pedrobaeza and @chienandalu can you review it?

@Tecnativa TT40363

### CODE REVIEW

File: project_stock/models/stock_move.py [hunk 1]
```diff
@@ -45,6 +46,7 @@ def _prepare_analytic_line_from_task(self):
             "product_uom_id": self.product_uom.id,
             "company_id": analytic_account.company_id.id or self.env.company.id,
             "partner_id": task.partner_id.id or task.project_id.partner_id.id or False,
+            "project_id": task.project_id.id,
```
Comment (pedrobaeza) [inline project_stock/models/stock_move.py]: This was removed on purpose, as only timesheets have the project filled.

File: project_stock/models/stock_move.py [hunk 2]
```diff
@@ -60,10 +62,10 @@ def _prepare_analytic_line_from_task(self):
         if "product_id" in analytic_line_fields:
             vals["product_id"] = product.id
         # Extra field added in hr_timesheet addon
-        if "employee_id" in analytic_line_fields:
+        if "employee_id" in analytic_line_fields and task.user_ids:
```
Comment (pedrobaeza) [inline project_stock/models/stock_move.py]: This shouldn't be filled as well. These are not timesheets.

File: project_stock/tests/test_project_stock.py [hunk 3]
```diff
@@ -51,121 +58,130 @@ def test_project_task_misc(self):
...
+            task.stock_analytic_tag_ids,
+            task.stock_analytic_line_ids.mapped("tag_ids"),
         )
         self.assertIn(
             self.analytic_account,
-            self.task.stock_analytic_line_ids.mapped("account_id"),
+            task.stock_analytic_line_ids.mapped("account_id"),
         )
 
+    @users("manager-user")
```
Comment (pedrobaeza) [inline project_stock/tests/test_project_stock.py]: Do both with admin and this user

File: project_stock/tests/test_project_stock.py [hunk 4]
```diff
@@ -51,121 +58,130 @@ def test_project_task_misc(self):
...
 
     def _test_task_analytic_lines_from_task(self, amount):
-        self.assertEqual(len(self.task.stock_analytic_line_ids), 2)
-        self.assertEqual(
-            sum(self.task.stock_analytic_line_ids.mapped("unit_amount")), 3
-        )
-        self.assertEqual(
-            sum(self.task.stock_analytic_line_ids.mapped("amount")), amount
-        )
+        task = self.env["project.task"].browse(self.task.id)
```
Comment (pedrobaeza) [inline project_stock/tests/test_project_stock.py]: For doing the diff less bigger, just do:

```suggestion
        self.task = self.env["project.task"].browse(self.task.id)
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: project_stock/models/stock_move.py (hunk 1 above)
Comment (victoralmau) [reply project_stock/models/stock_move.py]: Reviewing v14, I see that it auto-defined `project_id` with `hr_timesheet` (not sure how yet).

Comment (pedrobaeza) [reply project_stock/models/stock_move.py]: But these are not timesheets. That's why you shouldn't set it.

File: project_stock/tests/test_project_stock.py (hunk 3 above)
Comment (victoralmau) [reply project_stock/tests/test_project_stock.py]: Ok, I have added admin user in tests, but currently it is not possible to add multiple users with the `@users()` decorator without having side effect (no roolback of changes per user).

Comment (pedrobaeza) [reply project_stock/tests/test_project_stock.py]: What about this?: Create another method `test_..._admin` with the other @users decorator and calling the first method

**Comment by victoralmau** [comment]
Changes done and tests fixed.

File: project_stock/models/project_task.py [hunk 5]
```diff
@@ -74,6 +74,9 @@ class ProjectTask(models.Model):
     )
     stock_analytic_tag_ids = fields.Many2many(
         comodel_name="account.analytic.tag",
+        relation="account_analytic_tag_project_task_stock_rel",
```
Comment (pedrobaeza) [inline project_stock/models/project_task.py:77]: This will require migration scripts?

File: project_stock/security/project_stock_security.xml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<odoo>
+    <record id="account_analytic_line_project_stock_rule" model="ir.rule">
+        <field name="name">account.analytic.line.project_stock</field>
+        <field name="model_id" ref="analytic.model_account_analytic_line" />
+        <field name="domain_force">[('stock_task_id', '!=', False)]</field>
+        <field
+            name="groups"
+            eval="[(4, ref('analytic.group_analytic_accounting')), (4, ref('project.group_project_manager'))]"
```
Comment (pedrobaeza) [inline project_stock/security/project_stock_security.xml]: Why is this needed? This should be bypassed other way (with `sudo` for example).

File: project_stock/tests/test_project_stock.py [hunk 7]
```diff
@@ -76,7 +93,16 @@ def test_project_task_without_analytic_account(self):
         self.task.action_done()
         self.assertFalse(self.task.stock_analytic_line_ids)
 
-    def test_project_task_analytic_lines_without_tags(self):
+    @users("admin")
```
Comment (pedrobaeza) [inline project_stock/tests/test_project_stock.py]: No need of this decorator if you keep the other method without renaming nor users decorator (it already run with admin), and only add the other method.

File: project_stock/models/project_task.py (hunk 5 above)
Comment (victoralmau) [reply project_stock/models/project_task.py:77]: Yes, it requires renaming the table in migration script (already done).

File: project_stock/security/project_stock_security.xml (hunk 6 above)
Comment (victoralmau) [reply project_stock/security/project_stock_security.xml]: This is necessary because if `hr_timesheet` addon is installed, the analytic lines generated according to the rules cannot be accessed: https://github.com/odoo/odoo/blob/15.0/addons/hr_timesheet/security/hr_timesheet_security.xml

Comment (pedrobaeza) [reply project_stock/security/project_stock_security.xml]: And that's the idea... HR timesheet users only see timesheet entries, not these ones. Only accounting managers with full access will see all the analytic entries.

Comment (victoralmau) [reply project_stock/security/project_stock_security.xml]: Ok, removed

**Review by pedrobaeza — APPROVED** [review]
Code review

File: project_stock/README.rst [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    * `Type of operation`: Delivery
+    * `Default Source Location`: WH/Stock
+    * `Default Destination Location`: WH/Stock/Shelf 1
+#. Go to *Project -> Configuration -> Projects*.
+#. Create a new project with the following options:
+    * `Name`: Task material
+    * `Operation type`: Task material
+#. Go to *Project -> Configuration -> Task Stages* and edit some records.
+    * `In progress`: Check Use Stock Moves option and add the created project.
+    * `Done`: Check Use Stock Moves option + Lock Stock Moves and add the created project.
```
Comment (chienandalu) [inline project_stock/README.rst]:
```suggestion
    * `Done`: Check Use Stock Moves option + Done Stock Moves and add the created project.
```

File: project_stock/README.rst [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+Usage
+=====
+
+#. Go to *Projects -> Task material (project)* and create a task and edit it.
+#. *Stock Info* is displayed in the *Extra info* tab with the same project information, but it can be modified.
+#. Add some product to *Stock Info* tab and set some initial demand (1 for example).
+#. Change the stage to Done.
+#. *Stock Info* tab is readonly and some buttons show in header:
+    * `Confirm material`: Confirm all moves.
```
Comment (chienandalu) [inline project_stock/README.rst]: This one show up only in the "In progress" stage

**Review by chienandalu — CHANGES_REQUESTED** [review]
I detected an issue in my functional test:

- When I try to add more than one product line to the task and save it the following error is raised:

```
UncaughtPromiseError > TypeError
Uncaught Promise > Cannot read properties of undefined (reading 'type')

TypeError: Cannot read properties of undefined (reading 'type')
    at compareRecords (http://localhost:15069/web/assets/debug/web.assets_backend.js:69576:45)
    at stableCompare (http://localhost:15069/web/assets/debug/web.assets_common.js:66639:25)
    at Array.sort (<anonymous>) (/web/static/src/legacy/js/views/basic/basic_model.js:5170)
    at Object.stableSort (http://localhost:15069/web/assets/debug/web.assets_common.js:66638:22) (/web/static/src/legacy/js/core/utils.js:904)
    at Class._sortList (http://localhost:15069/web/assets/debug/web.assets_backend.js:69588:19) (/web/static/src/legacy/js/core/utils.js:903)
    at http://localhost:15069/web/assets/debug/web.assets_backend.js:69269:26 (/web/static/src/legacy/js/views/basic/basic_model.js:5182)
    at async Promise.all (index 4) (/web/static/src/legacy/js/views/basic/basic_model.js:4863)
    at async Promise.all (index 0)

**Comment by victoralmau** [comment]
> I detected an issue in my functional test:
> 
>     * When I try to add more than one product line to the task and save it the following error is raised:

Changes done and and problem solved.

**Review by chienandalu — APPROVED** [review]
Now the test went perfect :+1: Good job!
